radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java

```diff
@@ -225,6 +250,34 @@ public LedgerEntryStoreResult store(
 		}, CounterType.ELAPSED_BDB_LEDGER_LAST_VERTEX, CounterType.COUNT_BDB_LEDGER_LAST_VERTEX);
 	}
 
+	@Override
+	public void forEach(Consumer<SpunParticle> particleConsumer) {
+		try (var cursor = particleDatabase.openCursor(null, null)) {
+			var key = entry();
+			var value = entry();
+			var status = cursor.getFirst(key, value, DEFAULT);
+
+			while (status == SUCCESS) {
+				if (value.getData().length > 0) {
+					var particleBytes = Arrays.copyOfRange(value.getData(), EUID.BYTES, value.getData().length);
+					var particle = deserializeOrElseFail(particleBytes, Particle.class);
+					particleConsumer.accept(SpunParticle.up(particle));
+				} else {
+					//TODO: eventually we may want to store down particles as well
+					particleConsumer.accept(SpunParticle.down(VOID_PARTICLE));
```

Comment:
No won't be storing down particles in this part of the DB. Balance should only be dependent on UP particles so this is not needed.

Reply:
To calculate balances properly I need to have both, UP and DOWN particles. Also, beside balances there is a need to calculate token supply. The only method I see here is to handle unallocated tokens particle. Again, both, UP and DOWN particles are necessary to calculate supply.

Actually main issue here is the retrieving historical data during DB rebuild. Under normal circumstances full body of UP and DOWN particles are available, so there is no problem to calculate all things mentioned above.